### PR TITLE
feat: user-defined metadata support

### LINKED ISSUE #19: User defined metadata
Adding support for [user-defined object metadata](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html#UserMetadata) would be nice and not too hard to do, considering it is only a simple key value record added to the headers prefixed with `x-amz-meta-`.

### PR DESCRIPTION
**Pull request by eliassjogreen**
closes #19

### CODE REVIEW

File: deps.ts [hunk 1]
```diff
@@ -7,12 +7,12 @@ export type {
...
 } from "https://deno.land/x/aws_sign_v4@0.1.5/mod.ts";
-import { createHash } from "https://deno.land/std@0.79.0/hash/mod.ts";
+import { createHash } from "https://deno.land/std@0.83.0/hash/mod.ts";
 export function sha256Hex(data: string | Uint8Array): string {
   const hasher = createHash("sha256");
   hasher.update(data);
   return hasher.toString("hex");
 }
-export { default as parseXML } from "https://raw.githubusercontent.com/nekobato/deno-xml-parser/0bc4c2bd2f5fad36d274279978ca57eec57c680c/index.ts";
+export { default as parseXML } from "https://denopkg.com/nekobato/deno-xml-parser/index.ts";
```
Comment (lucacasonato) [inline deps.ts]: can you revert this?

File: src/bucket.ts [hunk 2]
```diff
@@ -258,6 +273,7 @@ export class S3Bucket {
       versionId: res.headers.get("x-amz-version-id") ?? undefined,
       websiteRedirectLocation:
         res.headers.get("x-amz-website-redirect-location") ?? undefined,
+      meta: Object.keys(meta).length !== 0 ? meta : undefined,
```
Comment (lucacasonato) [inline src/bucket.ts]: change this to always be set to meta (non optional). If no meta keys are set, this should be an empty object.

File: src/types.ts [hunk 3]
```diff
@@ -198,6 +198,12 @@ export interface HeadObjectResponse {
    */
   websiteRedirectLocation?: string;
 
+  /**
+   * The user-defined metadata of this object as a number of key-value pairs
+   * with the prefixed `x-amz-meta-` stripped from the key
+   */
+  meta?: Record<string, string>;
```
Comment (lucacasonato) [inline src/types.ts]: Make non optional.

**Review by lucacasonato — COMMENTED** [review]
Sorry that I didn't get to this! I totally missed the GitHub email. Looks good, just some minor comments.

File: src/types.ts [hunk 4]
```diff
@@ -390,7 +390,7 @@ export interface PutObjectOptions {
    * User-defined object metadata passed as a number of key-value pairs
    * that are converted into `x-amz-meta-` prefixed key headers.
    */
-  meta?: Record<string, string>;
+  meta: Record<string, string>;
```
Comment (lucacasonato) [inline src/types.ts]: You made it non optional on the `PutObjectOptions`, not the `HeadObjectOptions` :-)

File: src/bucket.ts [hunk 5]
```diff
@@ -273,7 +273,7 @@ export class S3Bucket {
       versionId: res.headers.get("x-amz-version-id") ?? undefined,
       websiteRedirectLocation:
         res.headers.get("x-amz-website-redirect-location") ?? undefined,
-      meta: Object.keys(meta).length !== 0 ? meta : undefined,
+      meta: Object.keys(meta).length !== 0 ? meta : {},
```
Comment (lucacasonato) [inline src/bucket.ts]: Is the ternary even needed here? Doesn't `Object.keys(meta).length == 0` mean that it is a `{}`?

Comment (eliassjogreen) [reply src/bucket.ts]: very true

File: src/types.ts (hunk 4 above)
Comment (eliassjogreen) [reply src/types.ts]: oops

**Review by lucacasonato — APPROVED** [review]
LGTM, thanks. I'll tag a new release soon.

**Pull request merged**